File: src/services/feed.service.ts
import type { PrismaClient } from '@prisma/client';
import {
  type CreateFeedInput,
  FeedRepository,
  type FeedWithFilters,
} from '../database/repositories/feed.repository.js';
import { feedQueueService } from '../jobs/index.js';
import { ConverterService } from '../utils/converters/converter.service.js';
import { logger } from '../utils/logger/logger.service.js';
import { isValidUrl } from '../utils/validation.js';
import { classifySource } from '../utils/source-classifier.js';
import { FeedDiscovery } from '../utils/feed-discovery.js';
import { providerRegistry } from '../providers/index.js';

export interface AddFeedInput {
  chatId: string;
  name: string;
  url: string;
  rssUrl?: string;
  enabled?: boolean;
}

export interface FeedValidationError {
  field: string;
  message: string;
}

export class FeedService {
  private feedRepository: FeedRepository;
  private converterService: ConverterService;

  constructor(prisma: PrismaClient) {
    this.feedRepository = new FeedRepository(prisma);
    this.converterService = new ConverterService();
  }

  /**
   * Add a new feed with validation and duplicate checking
   */
  async addFeed(input: AddFeedInput): Promise<{
    success: boolean;
    feed?: FeedWithFilters;
    errors?: FeedValidationError[];
    conversionInfo?: { originalUrl: string; rssUrl: string; platform?: string };
  }> {
    // Check for duplicate feed name and URL first
    const existingFeeds = await this.listFeeds(input.chatId);
    const duplicateName = existingFeeds.find(feed => feed.name.toLowerCase() === input.name.toLowerCase());
    const duplicateUrl = existingFeeds.find(feed => feed.url === input.url || feed.rssUrl === input.url);
    
    if (duplicateName) {
      return {
        success: false,
        errors: [{ field: 'name', message: `A feed with the name "${input.name}" already exists. Use a different name.` }],
      };
    }
    
    if (duplicateUrl) {
      return {
        success: false,
        errors: [{ field: 'url', message: `A feed with this URL already exists: "${input.url}". Check your feed list.` }],
      };
    }

    // Check global feed limit (100 feeds across all chats)
    const totalFeedsCount = await this.feedRepository.countAll();
    if (totalFeedsCount >= 100) {
      return {
        success: false,
        errors: [{ 
          field: 'limit', 
          message: `System limit reached: ${totalFeedsCount}/100 feeds. Please remove unused feeds before adding new ones.` 
        }],
      };
    }

    try {
      // Validate input
      const validationErrors = await this.validateFeedInput(input);
      if (validationErrors.length > 0) {
        return { success: false, errors: validationErrors };
      }

      // Classify source type (Reddit, RSS, etc.)
      const sourceType = classifySource(input.url);
      
      // Determine RSS URL through conversion or direct use
      let rssUrl = input.rssUrl || input.url;
      let conversionInfo: { originalUrl: string; rssUrl: string; platform?: string } | undefined;

      // If no explicit RSS URL provided, try URL conversion and feed discovery
      if (!input.rssUrl) {
        // For Reddit URLs, always use the URL as-is (our parser will handle it)
        if (sourceType === 'reddit') {
          rssUrl = input.url;
          logger.info(`Reddit URL detected: ${input.url}`);
          conversionInfo = {
            originalUrl: input.url,
            rssUrl: input.url,
            platform: 'reddit',
          };
        } else if (this.converterService.isRssUrl(input.url)) {
          // Check if URL is already in RSS format
          rssUrl = input.url;
          logger.info(`URL is already in RSS format: ${input.url}`);
        } else {
          // First try social media provider (Instagram, etc.)
          const bridgeUrl = providerRegistry.buildFeedUrl(input.url);
          if (bridgeUrl) {
            rssUrl = bridgeUrl;
            const provider = providerRegistry.findProvider(input.url);
            conversionInfo = {
              originalUrl: input.url,
              rssUrl: bridgeUrl,
              platform: provider?.name || 'bridge',
            };
            logger.info(`URL converted via bridge provider: ${input.url} -> ${bridgeUrl}`);
          } else {
            // Try traditional URL conversion (for YouTube, etc.)
            const conversionResult = await this.converterService.convertUrl(input.url);

            if (conversionResult.success && conversionResult.rssUrl) {
              rssUrl = conversionResult.rssUrl;
              conversionInfo = {
                originalUrl: input.url,
                rssUrl: conversionResult.rssUrl,
                platform: conversionResult.platform,
              };
              logger.info(
                `URL converted successfully: ${input.url} -> ${rssUrl} (${conversionResult.platform})`
              );
            } else {
              // Conversion failed, try feed discovery
              logger.info(`URL conversion failed, trying feed discovery for: ${input.url}`);
              const discoveryResult = await this.discoverFeeds(input.url);

              if (discoveryResult.success && discoveryResult.feeds.length > 0) {
                // Use the best feed found
                const bestFeed = discoveryResult.feeds[0];
                if (bestFeed) {
                  // Check if the discovered feed URL is already in use
                  const currentFeeds = await this.listFeeds(input.chatId);
                  const duplicateDiscoveredUrl = currentFeeds.find(feed => 
                    feed.url === bestFeed.url || feed.rssUrl === bestFeed.url
                  );
                  
                  if (duplicateDiscoveredUrl) {
                    return {
                      success: false,
                      errors: [{ 
                        field: 'url', 
                        message: `O feed descoberto "${bestFeed.url}" já está sendo usado pelo feed "${duplicateDiscoveredUrl.name}".` 
                      }],
                    };
                  }
                  
                  rssUrl = bestFeed.url;
                  conversionInfo = {
                    originalUrl: input.url,
                    rssUrl: bestFeed.url,
                    platform: `discovered-${bestFeed.source}`,
                  };
                  logger.info(
                    `Feed discovered: ${input.url} -> ${rssUrl} (${bestFeed.type}, confidence: ${bestFeed.confidence})`
                  );
                }
              } else {
                // Both conversion and discovery failed
                logger.warn(
                  `Both URL conversion and feed discovery failed for ${input.url}. Attempting to use original URL.`
                );
                return {
                  success: false,
                  errors: [{ 
                    field: 'url', 
                    message: `😤 Feed not found. Unable to discover or convert feed from "${input.url}". Please verify the URL has an RSS/Atom feed.` 
                  }],
                };
              }
            }
          }
        }
      }

      // Determine defaults based on feed type
      const isReddit = rssUrl.includes('reddit.com');
      const isInstagram = input.url.includes('instagram.com');
      
      // Get polling interval from provider if available
      const providerInterval = providerRegistry.getPollInterval(input.url);
      const checkIntervalMinutes = providerInterval || (isReddit ? 6 : 10);
      
      // Determine max age based on feed type
      const maxAgeMinutes = isReddit ? 90 : isInstagram ? 90 : 1440; // Reddit/IG: 90 min, others: 24h
      
      // Create feed
      const feedData: CreateFeedInput = {
        chat: {
          connect: { id: input.chatId },
        },
        name: input.name,
        url: input.url,
        rssUrl: rssUrl,
        enabled: input.enabled ?? true,
        checkIntervalMinutes,
        maxAgeMinutes,
        lastCheck: new Date(),
        failures: 0,
      };

      const feed = await this.feedRepository.create(feedData);
      logger.info(`Feed added successfully: ${feed.name} (${feed.id})`);

      // Schedule recurring feed checks using feed's checkIntervalMinutes
      try {
        const intervalMinutes = feed.checkIntervalMinutes;
        
        await feedQueueService.scheduleRecurringFeedCheck({
          feedId: feed.id,
          chatId: feed.chatId,
          feedUrl: feed.rssUrl,
          lastItemId: feed.lastItemId ?? undefined,
        }, intervalMinutes);

        logger.info(`Scheduled recurring feed check for feed ${feed.id} every ${intervalMinutes} minutes`);
      } catch (error) {
        logger.error(`Failed to schedule feed check for feed ${feed.id}:`, error);
        // Don't fail the entire operation if scheduling fails
      }

      return {
        success: true,
        feed,
        conversionInfo,
      };
    } catch (error) {
      logger.error('Error adding feed:', error);
      return {
        success: false,
        errors: [{ field: 'general', message: 'Failed to add feed due to internal error' }],
      };
    }
  }

  /**
   * List feeds for a specific chat
   */
  async listFeeds(chatId: string): Promise<FeedWithFilters[]> {
    try {
      return await this.feedRepository.findByChatId(chatId);
    } catch (error) {
      logger.error('Error listing feeds:', error);
      return [];
    }
  }

  /**
   * Remove a feed by name from a specific chat
   */
  async removeFeed(chatId: string, name: string): Promise<{ success: boolean; message: string }> {
    try {
      const feed = await this.feedRepository.findByChatIdAndName(chatId, name);
      if (!feed) {
        return { success: false, message: 'Feed not found' };
      }

      logger.info(`Starting transactional removal of feed: ${name} (${feed.id})`);

      // Use database transaction to ensure atomicity
      await this.feedRepository['prisma'].$transaction(async (tx: any) => {
        // Step 1: Remove recurring feed check job from queue first
        logger.info(`Removing recurring job for feed ${feed.id}`);
        const jobRemoved = await feedQueueService.removeRecurringFeedCheck(feed.id);
        
        if (!jobRemoved) {
          logger.error(`Failed to remove recurring job for feed ${feed.id}, aborting transaction`);
          throw new Error(`Failed to remove recurring job for feed ${feed.id}. Feed deletion aborted to prevent orphaned jobs.`);
        }

        // Step 2: Verify job removal was successful
        const jobVerified = await feedQueueService.verifyJobRemoval(feed.id);
        if (!jobVerified) {
          logger.error(`Job removal verification failed for feed ${feed.id}, aborting transaction`);
          throw new Error(`Job removal verification failed for feed ${feed.id}. Feed deletion aborted.`);
        }

        // Step 3: Delete feed record from database
        logger.info(`Deleting feed record for ${feed.id}`);
        await tx.feed.delete({
          where: { id: feed.id }
        });

        logger.info(`✅ Successfully completed transactional removal of feed: ${name} (${feed.id})`);
      });

      logger.info(`Feed removed successfully: ${name} (${feed.id})`);
      return { success: true, message: 'Feed removed successfully' };

    } catch (error) {
      logger.error(`Error removing feed ${name}:`, error);
      
      // Provide specific error messages based on the failure type
      const errorMessage = error instanceof Error ? error.message : 'Unknown error occurred';
      
      if (errorMessage.includes('Failed to remove recurring job')) {
        return { 
          success: false, 
          message: 'Failed to remove feed: Could not clean up scheduled jobs. Please try again or contact support.' 
        };
      }
      
      if (errorMessage.includes('Job removal verification failed')) {
        return { 
          success: false, 
          message: 'Failed to remove feed: Job cleanup verification failed. Please try again.' 
        };
      }

      return { 
        success: false, 
        message: 'Failed to remove feed due to internal error. Please try again.' 
      };
    }
  }

  /**
   * Enable a feed
   */
  async enableFeed(
    chatId: string,
    name: string
  ): Promise<{ success: boolean; message: string; feed?: FeedWithFilters }> {
    try {
      const feed = await this.feedRepository.findByChatIdAndName(chatId, name);
      if (!feed) {
        return { success: false, message: 'Feed not found' };
      }

      if (feed.enabled) {
        return { success: false, message: 'Feed is already enabled' };
      }

      const updatedFeed = await this.feedRepository.toggleEnabled(feed.id, true);
      logger.info(`Feed enabled successfully: ${name} (${feed.id})`);

      return { success: true, message: 'Feed enabled successfully', feed: updatedFeed };
    } catch (error) {
      logger.error('Error enabling feed:', error);
      return { success: false, message: 'Failed to enable feed due to internal error' };
    }
  }

  /**
   * Disable a feed
   */
  async disableFeed(
    chatId: string,
    name: string
  ): Promise<{ success: boolean; message: string; feed?: FeedWithFilters }> {
    try {
      const feed = await this.feedRepository.findByChatIdAndName(chatId, name);
      if (!feed) {
        return { success: false, message: 'Feed not found' };
      }

      if (!feed.enabled) {
        return { success: false, message: 'Feed is already disabled' };
      }

      const updatedFeed = await this.feedRepository.toggleEnabled(feed.id, false);
      logger.info(`Feed disabled successfully: ${name} (${feed.id})`);

      return { success: true, message: 'Feed disabled successfully', feed: updatedFeed };
    } catch (error) {
      logger.error('Error disabling feed:', error);
      return { success: false, message: 'Failed to disable feed due to internal error' };
    }
  }

  /**
   * Get feed count for a chat
   */
  async getFeedCount(chatId: string): Promise<number> {
    try {
      return await this.feedRepository.countByChatId(chatId);
    } catch (error) {
      logger.error('Error getting feed count:', error);
      return 0;
    }
  }

  /**
   * Get the converter service instance for external use
   */
  getConverterService(): ConverterService {
    return this.converterService;
  }

  /**
   * Discover available feeds from a website URL
   */
  async discoverFeeds(websiteUrl: string): Promise<{
    success: boolean;
    feeds: Array<{
      url: string;
      type: string;
      title?: string;
      confidence: number;
      source: string;
    }>;
    errors: string[];
  }> {
    try {
      logger.info(`Starting feed discovery for website: ${websiteUrl}`);

      const discoveryResult = await FeedDiscovery.discoverFeeds(websiteUrl);

      if (!discoveryResult.success) {
        return {
          success: false,
          feeds: [],
          errors: discoveryResult.errors,
        };
      }

      // Convert to our format
      const feeds = discoveryResult.feeds.map(feed => ({
        url: feed.url,
        type: feed.type,
        title: feed.title,
        confidence: feed.confidence,
        source: feed.source,
      }));

      logger.info(`Feed discovery completed: found ${feeds.length} feeds for ${websiteUrl}`);

      return {
        success: true,
        feeds,
        errors: discoveryResult.errors,
      };
    } catch (error) {
      logger.error('Error during feed discovery:', error);
      return {
        success: false,
        feeds: [],
        errors: [error instanceof Error ? error.message : 'Unknown error during discovery'],
      };
    }
  }


  /**
   * Check if a URL can be converted and preview the conversion
   */
  async previewUrlConversion(
    url: string
  ): Promise<{ canConvert: boolean; platform?: string; rssUrl?: string; error?: string }> {
    try {
      if (!isValidUrl(url)) {
        return { canConvert: false, error: 'Invalid URL format' };
      }

      if (this.converterService.isRssUrl(url)) {
        return { canConvert: false, error: 'URL is already in RSS format' };
      }

      const platform = this.converterService.detectPlatform(url);
      if (!platform) {
        return { canConvert: false, error: 'No converter available for this URL type' };
      }

      const conversionResult = await this.converterService.convertUrl(url);

      if (conversionResult.success && conversionResult.rssUrl) {
        return {
          canConvert: true,
          platform: conversionResult.platform,
          rssUrl: conversionResult.rssUrl,
        };
      }
      return {
        canConvert: false,
        platform,
        error: conversionResult.error || 'Conversion failed',
      };
    } catch (error) {
      logger.error('Error previewing URL conversion:', error);
      return {
        canConvert: false,
        error: 'Failed to preview conversion due to internal error',
      };
    }
  }

  /**
   * Validate feed input data
   */
  private async validateFeedInput(input: AddFeedInput): Promise<FeedValidationError[]> {
    const errors: FeedValidationError[] = [];

    // Validate name
    if (!input.name || input.name.trim().length === 0) {
      errors.push({ field: 'name', message: 'Feed name is required' });
    } else if (input.name.length > 100) {
      errors.push({ field: 'name', message: 'Feed name must be less than 100 characters' });
    }

    // Validate URL
    if (!input.url || input.url.trim().length === 0) {
      errors.push({ field: 'url', message: 'Feed URL is required' });
    } else if (!isValidUrl(input.url)) {
      errors.push({ field: 'url', message: 'Invalid URL format' });
    }

    // Validate chatId
    if (!input.chatId || input.chatId.trim().length === 0) {
      errors.push({ field: 'chatId', message: 'Chat ID is required' });
    }

    return errors;
  }
}

File: app/services/feed_service.py
"""Feed service for managing feeds"""

from typing import Optional, Dict, Any, List
from datetime import datetime
from uuid import uuid4
from sqlmodel import select

from app.database import database
from app.models.feed import Feed, Chat
from app.services.rss_service import rss_service
from app.utils.logger import get_logger

logger = get_logger(__name__)


class FeedService:
    """Service for managing feeds"""

    async def list_feeds(self, chat_id: str) -> List[Feed]:
        """List all feeds for a chat"""
        with database.get_session() as session:
            statement = select(Feed).where(Feed.chat_id == chat_id)
            feeds = session.exec(statement).all()
            return list(feeds)

    async def get_feed(self, chat_id: str, name: str) -> Optional[Feed]:
        """Get a specific feed"""
        with database.get_session() as session:
            statement = select(Feed).where(Feed.chat_id == chat_id, Feed.name == name)
            return session.exec(statement).first()

    async def add_feed(self, chat_id: str, name: str, url: str) -> Dict[str, Any]:
        """Add a new feed"""
        try:
            feed_url = url

            # Convert YouTube URLs to RSS format BEFORE validation
            from app.services.youtube_service import youtube_service

            if youtube_service.is_youtube_url(url):
                rss_url = youtube_service.convert_to_rss_url(url)
                if rss_url:
                    feed_url = rss_url
                    logger.info(f"Converting YouTube URL to RSS: {url} -> {feed_url}")
                else:
                    return {
                        "success": False,
                        "error": "Could not convert YouTube URL to RSS feed",
                    }

            # Convert Reddit URLs to RSS format BEFORE validation
            if feed_url == url:  # Only check Reddit if not already converted
                from app.services.reddit_service import reddit_service

                if reddit_service.is_reddit_url(url):
                    # Convert to RSS URL if not already in RSS format
                    if not url.endswith(".rss") and not url.endswith(".xml"):
                        if "/r/" in url:
                            subreddit = url.split("/r/")[1].split("/")[0]
                            feed_url = f"https://www.reddit.com/r/{subreddit}/.rss"
                            logger.info(f"Converting Reddit URL to RSS: {url} -> {feed_url}")
                        else:
                            feed_url = f"{url}.rss"

            # Validate feed URL (use converted URL for YouTube/Reddit)
            is_valid = await rss_service.validate_feed_url(feed_url)
            if not is_valid:
                return {
                    "success": False,
                    "error": "Invalid feed URL or feed is not accessible",
                }

            with database.get_session() as session:
                # Check if feed already exists
                existing = await self.get_feed(chat_id, name)
                if existing:
                    return {
                        "success": False,
                        "error": f"Feed '{name}' already exists",
                    }

                # Ensure chat exists
                chat = session.exec(select(Chat).where(Chat.id == chat_id)).first()
                if not chat:
                    chat = Chat(
                        id=chat_id,
                        type="private",  # Default, will be updated if needed
                        title=None,
                    )
                    session.add(chat)
                    session.commit()
                    session.refresh(chat)

                # Use the feed_url (already converted for Reddit if needed)
                rss_url = feed_url

                # Create feed
                feed = Feed(
                    id=str(uuid4()),
                    chat_id=chat_id,
                    name=name,
                    url=url,  # Keep original URL for display
                    rss_url=rss_url,  # Use converted RSS URL for fetching
                    check_interval_minutes=10,
                    max_age_minutes=1440,  # 24 hours
                    enabled=True,
                    failures=0,
                )
                session.add(feed)
                session.commit()
                session.refresh(feed)

                logger.info(f"Feed added: {chat_id}/{name}")
                return {"success": True, "feed": feed}

        except Exception as e:
            logger.error(f"Failed to add feed: {e}")
            return {"success": False, "error": str(e)}

    async def remove_feed(self, chat_id: str, name: str) -> Dict[str, Any]:
        """Remove a feed"""
        try:
            with database.get_session() as session:
                feed = await self.get_feed(chat_id, name)
                if not feed:
                    return {"success": False, "error": "Feed not found"}

                session.delete(feed)
                session.commit()

                logger.info(f"Feed removed: {chat_id}/{name}")
                return {"success": True}

        except Exception as e:
            logger.error(f"Failed to remove feed: {e}")
            return {"success": False, "error": str(e)}

    async def enable_feed(self, chat_id: str, name: str) -> Dict[str, Any]:
        """Enable a feed"""
        try:
            with database.get_session() as session:
                feed = await self.get_feed(chat_id, name)
                if not feed:
                    return {"success": False, "error": "Feed not found"}

                feed.enabled = True
                session.add(feed)
                session.commit()

                logger.info(f"Feed enabled: {chat_id}/{name}")
                return {"success": True}

        except Exception as e:
            logger.error(f"Failed to enable feed: {e}")
            return {"success": False, "error": str(e)}

    async def disable_feed(self, chat_id: str, name: str) -> Dict[str, Any]:
        """Disable a feed"""
        try:
            with database.get_session() as session:
                feed = await self.get_feed(chat_id, name)
                if not feed:
                    return {"success": False, "error": "Feed not found"}

                feed.enabled = False
                session.add(feed)
                session.commit()

                logger.info(f"Feed disabled: {chat_id}/{name}")
                return {"success": True}

        except Exception as e:
            logger.error(f"Failed to disable feed: {e}")
            return {"success": False, "error": str(e)}

    async def get_all_enabled_feeds(self) -> List[Feed]:
        """Get all enabled feeds"""
        try:
            with database.get_session() as session:
                # Get total feeds count for diagnostics
                total_statement = select(Feed)
                total_feeds = session.exec(total_statement).all()
                total_count = len(list(total_feeds))

                # Get enabled feeds
                statement = select(Feed).where(Feed.enabled)
                feeds = session.exec(statement).all()
                feeds_list = list(feeds)
                enabled_count = len(feeds_list)

                logger.info(
                    f"📈 Feed statistics: {enabled_count} enabled out of {total_count} total feeds"
                )

                if enabled_count == 0 and total_count > 0:
                    logger.warning(
                        f"⚠️ No enabled feeds found, but {total_count} feed(s) exist in database (they may be disabled)"
                    )
                elif total_count == 0:
                    logger.info("ℹ️ No feeds found in database at all")

                return feeds_list
        except Exception as e:
            logger.error(f"❌ Failed to get enabled feeds from database: {e}", exc_info=True)
            return []

    async def update_feed_last_check(
        self,
        feed_id: str,
        last_item_id: Optional[str] = None,
        last_notified_at: Optional[datetime] = None,
    ):
        """Update feed's last check time and last item ID"""
        try:
            with database.get_session() as session:
                feed = session.get(Feed, feed_id)
                if feed:
                    feed.last_check = datetime.utcnow()
                    if last_item_id:
                        feed.last_item_id = last_item_id
                        logger.debug(f"Updated feed {feed.name} lastItemId: {last_item_id}")
                    if last_notified_at:
                        feed.last_notified_at = last_notified_at
                        logger.debug(
                            f"Updated feed {feed.name} lastNotifiedAt: {last_notified_at.isoformat()}"
                        )
                    session.add(feed)
                    session.commit()
                    logger.debug(
                        f"Feed {feed.name} last check updated: {feed.last_check.isoformat()}"
                    )
        except Exception as e:
            logger.error(f"Failed to update feed last check: {e}")


# Global feed service instance
feed_service = FeedService()
